File: src/cnft/proof.ts
// Merkle proof computation for cNFT trees. Pure: given a TreeState with
// populated leaves and a leafIndex, emit the sibling path from leaf to
// root. Also exposes `verifyProof` so tests can cross-check.
//
// The algorithm is standard binary-merkle: at each level we emit the
// sibling hash, then fold (self, sibling) — or (sibling, self) depending
// on which side we're on — up to the parent, and repeat. Unpopulated
// positions use the empty-node cascade, so we don't need to store
// zero-filled subtrees.
//
// Runtime is O(depth * 2^depth_untouched) worst case if we naively
// recompute; we cut it to O(depth * filled_leaves) by folding only the
// populated positions per level. Good enough for trees up to depth 30
// with tens of thousands of leaves — above that a persistent node-level
// cache wins, but that's a later optimization.

import { emptyNode, hashPair } from "./hash.js";
import type { MerkleProof, TreeState } from "./types.js";

export function computeProof(
  tree: TreeState,
  leafIndex: bigint,
): MerkleProof {
  if (tree.depth < 1 || tree.depth > 30) {
    throw new Error(`computeProof: unsupported tree depth ${tree.depth}`);
  }
  const capacity = 1n << BigInt(tree.depth);
  if (leafIndex < 0n || leafIndex >= capacity) {
    throw new Error(
      `computeProof: leafIndex ${leafIndex} out of range for depth ${tree.depth} (capacity ${capacity})`,
    );
  }

  // Snapshot the current level as a map from position → node hash. Level
  // 0 is leaves; level `depth` is the root. Positions absent from the
  // map resolve to `emptyNode(level)`.
  let level = new Map<bigint, Uint8Array>(tree.leaves);
  const proof: Uint8Array[] = [];

  let currentIndex = leafIndex;
  for (let h = 0; h < tree.depth; h++) {
    const sibIdx = currentIndex ^ 1n;
    const sibling = level.get(sibIdx) ?? emptyNode(h);
    proof.push(sibling);

    // Fold this level into the next. We only need to compute nodes
    // whose children we've seen — anything else stays "empty" and is
    // derived on demand from emptyNode(h+1). Iterate unique parent
    // indexes that have at least one non-empty child.
    const next = new Map<bigint, Uint8Array>();
    const seenParents = new Set<bigint>();
    for (const pos of level.keys()) {
      const parent = pos >> 1n;
      if (seenParents.has(parent)) continue;
      seenParents.add(parent);
      const leftIdx = parent << 1n;
      const rightIdx = leftIdx + 1n;
      const left = level.get(leftIdx) ?? emptyNode(h);
      const right = level.get(rightIdx) ?? emptyNode(h);
      next.set(parent, hashPair(left, right));
    }
    level = next;
    currentIndex >>= 1n;
  }

  // After `depth` iterations, `level` contains exactly the root at
  // index 0 (or nothing, if the tree is entirely empty — in which case
  // the root is the empty cascade at `depth`).
  const root = level.get(0n) ?? emptyNode(tree.depth);

  const leaf = tree.leaves.get(leafIndex) ?? emptyNode(0);
  const nodeIndex = Number((1n << BigInt(tree.depth)) + leafIndex);

  return { leaf, proof, root, nodeIndex };
}

/**
 * Verify a proof bottom-up. `leafIndex` determines whether we hash
 * (self, sibling) or (sibling, self) at each level — the lowest bit
 * says which side of its parent the current node is on.
 */
export function verifyProof(args: {
  leaf: Uint8Array;
  proof: Uint8Array[];
  leafIndex: bigint;
  root: Uint8Array;
}): boolean {
  let current = args.leaf;
  let idx = args.leafIndex;
  for (const sibling of args.proof) {
    if ((idx & 1n) === 0n) {
      current = hashPair(current, sibling);
    } else {
      current = hashPair(sibling, current);
    }
    idx >>= 1n;
  }
  return eq(current, args.root);
}

function eq(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

File: crates/core/src/proof.rs
//! Merkle proof computation and verification for Tidepool cNFT trees.
//!
//! Standard binary-merkle: at each level, emit the sibling hash, fold
//! `(self, sibling)` or `(sibling, self)` depending on which side
//! we're on, repeat to the root. Unpopulated positions use the empty
//! cascade so we don't pay for zero-filled subtrees.
//!
//! Runtime is O(depth * populated_leaves). Good enough for depth-30
//! trees with tens of thousands of leaves. Above that, a persistent
//! node-level cache wins — future optimization, not v1 concern.

use std::collections::{BTreeMap, BTreeSet};

use crate::hash::{empty_node, hash_pair};
use crate::types::{MerkleProof, TreeState};

/// Errors compute_proof can surface. Out-of-band inputs (bad depth,
/// bad leaf index) — never triggered by on-chain data in practice, but
/// returned as `Result` rather than panicking so callers can handle
/// library-misuse cases gracefully.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ProofError {
    /// Tree depth must be between 1 and 30 inclusive.
    UnsupportedDepth(u8),
    /// Leaf index beyond the tree's capacity (`2^depth`).
    OutOfRange {
        leaf_index: u64,
        depth: u8,
        capacity: u64,
    },
}

impl std::fmt::Display for ProofError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnsupportedDepth(d) => {
                write!(f, "unsupported tree depth {d} (must be 1..=30)")
            }
            Self::OutOfRange {
                leaf_index,
                depth,
                capacity,
            } => write!(
                f,
                "leaf_index {leaf_index} out of range for depth {depth} (capacity {capacity})"
            ),
        }
    }
}

impl std::error::Error for ProofError {}

/// Compute a merkle proof for the leaf at `leaf_index` in `tree`.
///
/// Returns the proof path (sibling-of-leaf up to sibling-of-root), the
/// computed root, the leaf hash at that position (or empty-node(0) if
/// absent), and the node_index = `2^depth + leaf_index` used by DAS's
/// `getAssetProof` wire format.
pub fn compute_proof(tree: &TreeState, leaf_index: u64) -> Result<MerkleProof, ProofError> {
    if tree.depth < 1 || tree.depth > 30 {
        return Err(ProofError::UnsupportedDepth(tree.depth));
    }
    let depth = tree.depth;
    let capacity = 1u64 << depth;
    if leaf_index >= capacity {
        return Err(ProofError::OutOfRange {
            leaf_index,
            depth,
            capacity,
        });
    }

    // Working level: positions present at this height. Start at leaves.
    let mut level: BTreeMap<u64, [u8; 32]> = tree.leaves.clone();
    let mut proof: Vec<[u8; 32]> = Vec::with_capacity(depth as usize);
    let mut current_index = leaf_index;

    for h in 0..depth {
        let sib_idx = current_index ^ 1;
        let sibling = level
            .get(&sib_idx)
            .copied()
            .unwrap_or_else(|| empty_node(h as usize));
        proof.push(sibling);

        // Fold into the next level. Only compute parents whose
        // children include at least one populated position; everything
        // else is implicitly the empty cascade at the next height.
        let mut next: BTreeMap<u64, [u8; 32]> = BTreeMap::new();
        let mut seen: BTreeSet<u64> = BTreeSet::new();
        for &pos in level.keys() {
            let parent = pos >> 1;
            if !seen.insert(parent) {
                continue;
            }
            let left_idx = parent << 1;
            let right_idx = left_idx + 1;
            let left = level
                .get(&left_idx)
                .copied()
                .unwrap_or_else(|| empty_node(h as usize));
            let right = level
                .get(&right_idx)
                .copied()
                .unwrap_or_else(|| empty_node(h as usize));
            next.insert(parent, hash_pair(&left, &right));
        }
        level = next;
        current_index >>= 1;
    }

    // After `depth` folds, `level` has at most one entry at position 0.
    // If the tree was entirely empty, `level` is empty and the root
    // defaults to the empty cascade at `depth`.
    let root = level
        .get(&0)
        .copied()
        .unwrap_or_else(|| empty_node(depth as usize));

    let leaf = tree
        .leaves
        .get(&leaf_index)
        .copied()
        .unwrap_or_else(|| empty_node(0));

    let node_index = (1u64 << depth) + leaf_index;

    Ok(MerkleProof {
        leaf,
        proof,
        root,
        node_index,
    })
}

/// Verify a proof bottom-up. The lowest bit of `leaf_index` at each
/// level says which side of the parent the current node is on.
#[must_use]
pub fn verify_proof(leaf: &[u8; 32], proof: &[[u8; 32]], leaf_index: u64, root: &[u8; 32]) -> bool {
    let mut current = *leaf;
    let mut idx = leaf_index;
    for sibling in proof {
        current = if idx & 1 == 0 {
            hash_pair(&current, sibling)
        } else {
            hash_pair(sibling, &current)
        };
        idx >>= 1;
    }
    &current == root
}
